### DISCUSSION TITLE
Couple of questions

### DISCUSSION THREAD

**Discussion by segevfiner** [twpayne/chezmoi#2210] [Q&A] [answered]
1. What's the difference between `chezmoiscripta` and the `run_` attribute?
2. What's the difference between `run_` and `onchange_`? The docs weren't very clear about this.

**Comment by jfmercer** [comment]
I'm pretty new to chezmoi, but I believe that `run_` runs with every `chezmoi apply`, whereas `onchange_` runs with `chezmoi apply` only if the script has changed.

**Comment by bradenhilton** [comment, marked as answer]
> 1. What's the difference between `chezmoiscripta` and the `run_` attribute?

`.chezmoiscripts` is a special directory intended to help organize `run_*` scripts without creating a directory structure in the destination directory. For example, you can place all Linux `run_` scripts in `.chezmoiscripts/linux/`. Scrips which you intend to run on the machine outside of chezmoi (`executable_` scripts) should not be placed in this directory.

> 2. What's the difference between `run_` and `onchange_`? The docs weren't very clear about this.

From https://www.chezmoi.io/reference/target-types/#scripts:

> Scripts are executed on every `chezmoi apply`, unless they have the `once_` or `onchange_` attribute. `run_once_` scripts are only executed if a script with the same contents has not been run before, i.e. if the script is new or if its contents have changed. `run_onchange_` scripts are executed whenever their contents change, even if a script with the same contents has run before.

If I recall correctly:

- The contents of `run_once_` scripts are hashed and stored in chezmoi's cache.  The script will only be run again if its contents have changed.
- The filename of `run_onchange_` scripts, as well as the hash of their contents are stored in chezmoi's cache. The script will be run again if it's contents or name changes.

The cache is checked when running `chezmoi apply` to determine if the scripts have been run before.
